src/jit.c

```diff
@@ -0,0 +1,1557 @@
+        if (src == 4) /* SLL */
+            emit_dataproc_2source(state, false, DP2_LSLV, dst, dst, R8);
+        else if (src == 5) /* SRL */
+            emit_dataproc_2source(state, false, DP2_LSRV, dst, dst, R8);
+        else if (src == 7) /* SRA */
+            emit_dataproc_2source(state, false, DP2_ASRV, dst, dst, R8);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* REX prefix, ModRM byte, and 32-bit immediate */
+static inline void emit_alu32_imm32(struct jit_state *state,
+                                    int op UNUSED,
+                                    int src,
+                                    int dst,
+                                    int32_t imm)
+{
+#if defined(__x86_64__)
+    emit_alu32(state, op, src, dst);
+    emit4(state, imm);
+#elif defined(__aarch64__)
+    switch (src) {
+    case 0:
+        emit_load_imm(state, R10, imm);
+        emit_addsub_register(state, false, AS_ADD, dst, dst, R10);
+        break;
+    case 1:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_ORR, dst, dst, R10);
+        break;
+    case 4:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_AND, dst, dst, R10);
+        break;
+    case 6:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_EOR, dst, src, R10);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* REX prefix, ModRM byte, and 8-bit immediate */
+static inline void emit_alu32_imm8(struct jit_state *state,
+                                   int op UNUSED,
+                                   int src,
+                                   int dst,
+                                   int8_t imm)
+{
+#if defined(__x86_64__)
+    emit_alu32(state, op, src, dst);
+    emit1(state, imm);
+#elif defined(__aarch64__)
+    switch (src) {
+    case 4:
+        emit_load_imm(state, R10, imm);
+        emit_dataproc_2source(state, false, DP2_LSLV, dst, dst, R10);
+        break;
+    case 5:
+        emit_load_imm(state, R10, imm);
+        emit_dataproc_2source(state, false, DP2_LSRV, dst, dst, R10);
+        break;
+    case 7:
+        emit_load_imm(state, R10, imm);
+        emit_dataproc_2source(state, false, DP2_ASRV, dst, dst, R10);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* The REX.W prefix and ModRM byte are emitted.
+ * The MR encoding is used when there is a choice. 'src' is often used as
+ * an opcode extension.
+ */
+static inline void emit_alu64(struct jit_state *state, int op, int src, int dst)
+{
+#if defined(__x86_64__)
+    emit_basic_rex(state, 1, src, dst);
+    emit1(state, op);
+    emit_modrm_reg2reg(state, src, dst);
+#elif defined(__aarch64__)
+    if (op == 0x01)
+        emit_addsub_register(state, true, AS_ADD, dst, dst, src);
+#endif
+}
+
+/* REX.W prefix, ModRM byte, and 8-bit immediate */
```

Comment:
This comment should only make sense to x64 backend. Move into the function body.